File: backend/app/infrastructure/repositories/mappers/user_progress_mapper.py
from app.domain.entities.user_progress import UserProgress as UserProgressDomain
from app.infrastructure.repositories.orm.user_progress import UserProgress as UserProgressORM

def to_domain(orm_user_progress: UserProgressORM) -> UserProgressDomain:
    return UserProgressDomain(
        user_id=orm_user_progress.user_id,
        lesson_id=orm_user_progress.lesson_id,
        completed_at=orm_user_progress.completed_at
    )

File: frontend/lib/data/mappers/user_progress_mapper.dart
// frontend/lib/data/mappers/user_progress_mapper.dart
import 'package:frontend/data/models/user_progress_model.dart';
import 'package:frontend/domain/entities/user_progress.dart';
import 'package:frontend/data/mappers/lesson_mapper.dart'; // Assuming LessonMapper exists

class UserProgressMapper {
  static UserProgress fromModel(UserProgressModel model) {
    return UserProgress(
      userId: model.userId,
      lessonId: model.lessonId,
      completedAt: model.completedAt,
      lesson: model.lesson != null ? LessonMapper.fromModel(model.lesson!) : null, // Map nested Lesson if available
    );
  }

  static UserProgressModel toModel(UserProgress entity) {
    return UserProgressModel(
      userId: entity.userId,
      lessonId: entity.lessonId,
      completedAt: entity.completedAt,
      lesson: entity.lesson != null ? LessonMapper.toModel(entity.lesson!) : null, // Map nested Lesson if available
    );
  }
}
